File: 알고리즘실습/C++/lab7/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>

#define INF 0xffff
using namespace std;

typedef vector<vector<int>> matrix;
typedef struct edge_t{
    int u,v,w;
}edge_t;


typedef vector<edge_t> set_of_edges;

void printTable(int n, vector<int> table){
    if(n < 2){
        return;
    }
    cout << table[2];
    for(int i = 3; i <= n; i++)
        cout << ' ' << table[i];
    cout << '\n';
}

void printPath(vector<int> table, int n){
    if(n == 1){
        cout << 1 << '\n';
        return;
    }
    vector<int> path;
    int i;
    for(i = n; table[i] != 1; i = table[i])
        path.push_back(i);
    path.push_back(i);

    cout << 1;
    for(int j = path.size()-1; j >= 0; j--)
        cout << ' ' << path[j];
    cout << '\n';
}

vector<int> dijkstra(int n, matrix& W, set_of_edges& F){
    int vnear, min;
    vector<int> touch(n+1), length(n+1);
    F.clear();

    if(n == 1){
        printTable(n, touch);
        return touch;
    }

    for(int i = 2; i <= n; i++){
        touch[i] = 1;
        length[i] = W[1][i];
    }
    printTable(n, touch);

    for(int k = 0; k < n-1; k++){
        min = INF;
        vnear = -1;
        for(int i = 2; i <= n; i++)
            if(0 <= length[i] && length[i] < min){
                min = length[i];
                vnear = i;
            }
        if(vnear == -1) break;

        edge_t e;
        e.u = touch[vnear];
        e.v = vnear;
        e.w = W[touch[vnear]][vnear];  
        F.push_back(e);

        for(int i = 2; i <= n; i++)
            if(length[i] > length[vnear] + W[vnear][i]){
                length[i] = length[vnear] + W[vnear][i];
                touch[i] = vnear;
            }
        length[vnear] = -1;
        printTable(n, touch);
    }
    return touch;
}

int main(){
    int n, k, numOfTestCase;
    int u, v, w;
    cin >> n >> k;

    matrix W(n+1, vector<int>(n+1, INF));
    set_of_edges E, F;
    vector<int> testcase;

    for(int i = 0; i < k; i++){
        cin >> u >> v >> w;
        W[u][v] = w;       
        edge_t a;
        a.u = u;
        a.v = v;
        a.w = w;
        E.push_back(a);
    }

    cin >> numOfTestCase;
    for(int i = 0; i < numOfTestCase; i++){
        int tmp;
        cin >> tmp;
        testcase.push_back(tmp);
    }

    vector<int> touch = dijkstra(n, W, F);

    for(edge_t i : F)
        cout << i.u << ' ' << i.v << ' ' << i.w << '\n';

    for(int i : testcase)
        printPath(touch, i);

    return 0;
}

File: 알고리즘실습/python/dijkstra.py
INF = 0xffff

def print_table(n, table):
    if n < 2:
        print()
        return
    print(' '.join(str(table[i]) for i in range(2, n+1)))

def print_path(table, n):
    if n == 1:
        print(1)
        return
    path = []
    i = n
    while table[i] != 1:
        path.append(i)
        i = table[i]
    path.append(i)
    
    result = [1] + path[::-1]
    print(' '.join(map(str, result)))

def dijkstra(n, W):
    touch = [0] * (n + 1)
    length = [0] * (n + 1)
    F = []

    if n == 1:
        print_table(n, touch)
        return touch, F

    for i in range(2, n + 1):
        touch[i] = 1
        length[i] = W[1][i]
    print_table(n, touch)

    for _ in range(n - 1):
        min_val = INF
        vnear = -1
        for i in range(2, n + 1):
            if 0 <= length[i] < min_val:
                min_val = length[i]
                vnear = i
        if vnear == -1:
            break

        F.append((touch[vnear], vnear, min_val))

        for i in range(2, n + 1):
            if length[vnear] + W[vnear][i] < length[i]:
                length[i] = length[vnear] + W[vnear][i]
                touch[i] = vnear
        length[vnear] = -1
        print_table(n, touch)

    return touch, F

def main():
    n, k = map(int, input().split())

    W = [[INF] * (n + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        W[i][i] = 0

    for _ in range(k):
        u, v, w = map(int, input().split())
        W[u][v] = w

    c = int(input())
    testcases = [int(input()) for _ in range(c)]

    touch, F = dijkstra(n, W)

    for u, v, w in F:
        print(u, v, w)

    for t in testcases:
        print_path(touch, t)

if __name__ == "__main__":
    main()